STYLE_GUIDE.md

```diff
@@ -1,3 +1,66 @@
 # Fortran stdlib Style Guide
 
-This document will describe the code style to use.
+> [A]utomate and codify as much as you possibly can while remembering that the human touch is still necessary to praise
+> and be kind to your contributors.
+> Let robots handle your project’s pedantry and humans handle your project’s empathy.
+
+-- @mikemcquaid, [Homebrew] project leader<sup>[1]</sup>
+
+[1]: https://mikemcquaid.com/2018/06/05/robot-pedantry-human-empathy/
+[Homebrew]: https://brew.sh
+
+## File naming conventions
+
+- Source files should contain at most one `program` or `module`/`submodule`.
+- The filename should match the program or module name and have the file extension `.f90` or `.F90` if preprocessing is required.
+- If the interface and implementation is split using submodules the implementation submodule file should have the same name as the interface (parent) module but end in `_implementation`. E.g., `string_class.f90` and `string_class_implementation.f90`
+- Tests should be added in the `tests` subdirectory and have the same name as the module they are testing with the `test_` prefix added. E.g., `string_class.f90` and `tests/test_string_class.f90`
+
+## Indentation & whitespace
+
+By setting and following a convention for indentation and whitespace, code reviews and git-diffs can
+focus on the semantics of the proposed changes rather than style and formatting.
+
+* __TABs vs spaces__: __Spaces__: TABs should never be used for indentation
+* __Trailing whitespace__: All trailing whitespace must be removed
+* __End of File (EOF)__: The file must end in a single new-line character
+* __Line length__: __132__: The Fortran 90 (and later) standard mandates lines be no longer than 132 characters
+* __Indentation__: __Two (2) spaces__ for all control constructs, line continuations, etc.
+* __Line ending character__: __Native__: Git will convert line endings to `\n` (Unix style) on commit, and will checkout files with native line endings
+
+## Variable naming
+
+Variable names should be descriptive, and not artificially short.
+By default, where it makes sense to do so, variable names shall be made up of one or more full words separated by an underscore.
+For cases where a conventional & appropriate shortening of a word is used then the underscore may be omitted.
+
+Examples:
+
+__GOOD__:
+
+``` fortran
+logical :: has_failed
+real function linspace(...)
+```
+
+__BAD__:
+
+``` fortran
+logical :: has_failed
+real function lin_space(...)
+```
```

Comment:
(I can't seem to work out GitHub suggestion for multi-lines)

This whole thing can be condensed to:

* Variable names should be made up of one or more full words separated by an underscore, for example `has_failed` is preferred over `hasfailed`
* Where conventional and appropriate shortening of a word is used then the underscore may be omitted, for example `linspace` is preferred over `lin_space`

Reply:
I think `has_failed` and `lin_space` is equivalent. They could both be used, but if it's just two syllables, it is perfectly fine to remove the `_` and just make it `linspace` and `hasfailed`. It all depends on the context if one should be preferred over the other.

In general, it's worth not to be typing `_` all the time, and it actually somtimes improves readability to remove `_` for two syllable names. If it's more than two syllables, then the `_` improves readability.

Reply:
I see your point but I don't think they're quite equivalent. I see `linspace` is a programming portmanteau for "linear space". `linspace` makes more sense because `lin` is not a word on its own. `has_failed` however are two independent words, and like `is`, word `has` implies a logical state. I find `hasfailed` about as readable as the other one, but `has_*` has a semantic meaning implied in it (similar to `is_`, `get_`, `set_`, etc.).

Perhaps these are not the clearest example for the guide because the difference is nuanced.

Reply:
Yes, I agree that `is_alpha` as we discussed should be with `_` because `is` has a semantic meaning. In the same spirit, if that is the case with `has`, then it should be `has_failed`. (All I was saying is that I can imagine a context when `has` does not have a semantic meaning, in which case `hasfailed` works also.) So I wanted to keep our rules general enough to allow for both, based on the context.

Reply:
As a starting point I'm adopting @milancurcic's text/suggestion, but I'm happy to iterate on it more if someone can propose a better alternative that we agree upon.